File: 6/6_ParserBottonUp.js

let counter = 1;

class ParseTable {
  constructor(table) {
    this.Table = table;

    this.NotTerminals = [];
    let i = 1;
    while (i < table.length) {
      this.NotTerminals.push(table[i][0]);
      i++;
    }
    this.Headers = [];
    i = 1;

    while (i < table[0].length) {
      this.Headers.push(table[0][i]);
      i++;
    }
  }

  Print() {
    console.log("");
    for (let i = 0; i < this.Table.length; i++) {
      console.log(this.Table[i]);
    }
    console.log("");
  }
}

class NodeStack {
  constructor(symbol, terminal) {
    globalThis.counter;
    this.id = counter;
    this.symbol = symbol;
    this.terminal = terminal;
    counter += 1;
  }
}

class Token {
  constructor(type, lexeme, line) {
    this.type = type;
    this.lexeme = lexeme;
    this.line = line;
  }
}

const CreateTable = require('./read.js');

async function GetTable(fileName) {
    let table = [];
    try {
        table = await CreateTable(fileName);
    } 
    catch (error) {
        console.error(error);
    }
    return table;
}

function ReplaceStack(node, token, parseTable) {
    if (parseTable.NotTerminals.includes(node.symbol)) {
        const stack = [];
        const indexLex = parseTable.Headers.indexOf(token.lexeme);
        const indexSym = parseTable.NotTerminals.indexOf(node.symbol);

        console.log('indexLex: ',indexLex)
        console.log('indexSym: ',indexSym)

        console.log('token.lexeme: ',token.lexeme)
        console.log('node.symbol: ',node.symbol)

        let text = parseTable.Table[indexSym + 1][indexLex + 1];

        
        console.log('text: ',text)
        text = text.split(" ");
        console.log('text: ',text)

        for (let i = 0; i < text.length; i++) {
            if (parseTable.NotTerminals.includes(text[i])) {
                stack.push(new NodeStack(text[i], false));
            } 
            else {
                stack.push(new NodeStack(text[i], true));
            }
        }
        return stack;
    } 
    else {
        return [node];
    }
}

function PrintStack(Stack) {
    let symbols = Stack[0].symbol;
    for (let i = 1; i < Stack.length; i++) {
        symbols += Stack[i].symbol;
    }
    console.log("Stack: ", symbols);
}

function PrintTokens(Tokens) {
    let symbols = Tokens[0].lexeme;
    for (let i = 1; i < Tokens.length; i++) {
        symbols += Tokens[i].lexeme;
    }
    console.log("Tokens: ", symbols);
}

function Parser(stack, tokens, parseTable) {
    console.log();
    PrintStack(stack);
    PrintTokens(tokens);
    let i = 1;

    while (true) {
        if (stack[0].symbol === "$" && tokens[0].lexeme === "$") {
            console.log(true);
            break;
        }
        if (stack[0].symbol === tokens[0].lexeme && stack[0].terminal === true) {
            let pop = stack.shift().symbol;
            pop += " " + tokens.shift().lexeme;
        }
        if (stack[0].symbol === "ε") {
            let pop = stack.shift().symbol;
        }
        if (stack[0].symbol !== tokens[0].lexeme && stack[0].terminal === true) {
            console.log(false);
            break;
        }
        const node = stack.shift();
        stack = ReplaceStack(node, tokens[0], parseTable).concat(stack);
        i++;
    }
}

async function main() {
    try {
        let t1 = new Token("lparen", "(", 1);
        let t2 = new Token("int", "int", 1);
        let t3 = new Token("operator", "+", 1);
        let t4 = new Token("int", "int", 1);
        let t5 = new Token("rparen", ")", 1);
        let t6 = new Token("operator", "*", 1);
        let t7 = new Token("int", "int", 1);
        let te = new Token("$", "$", 1);

        let tokens = [t2,t6,t4,te];
        let tokens2 = [t1,t2,t3,t4,t5,te];
        let tokens3 = [t1,t2,t3,t4,t5,t6,t7,te];

        let node_E = new NodeStack("E", false);
        let node_dolar = new NodeStack("$", true);
        let stack = [ node_E, node_dolar ];

        var table = await GetTable('table.csv');
        //console.log('146: ',table);
        let parseTable = new ParseTable(table);

        Parser(stack.slice(),tokens,parseTable);
        Parser(stack.slice(),tokens2,parseTable);
        Parser(stack.slice(),tokens3,parseTable);

    }
    catch (error) {
        console.error(error);
    }
}
 
main();

File: 6/6_ParserBottonUp.py
import csv

counter = 1

class ParseTable:
  def __init__(self, table):
    self.Table = table
    self.NotTerminals = []
    i = 1
    while i < len(table):
      self.NotTerminals.append(table[i][0])
      i+=1
      
    self.Headers = []
    i = 1
    while i < len(table[0]):
      self.Headers.append(table[0][i])
      i+=1
      
  def Print(self):
    print('')
    for i in self.Table:
      print(i)
    print('')
    
class NodeStack:
  def __init__(self, symbol, terminal):
    global counter
    self.id = counter
    self.symbol = symbol
    self.terminal = terminal
    counter += 1

class Token:
  def __init__(self, type, lexeme, line):
    self.type = type
    self.lexeme = lexeme
    self.line = line

def CreateTable(file):
  with open(file, newline='') as csvfile:
    rows = csv.reader(csvfile, delimiter=',')
    table = [row for row in rows]
  return table

def ReplaceStack(node,token,parseTable):
  if node.symbol in parseTable.NotTerminals:
    stack = []
    indexLex = parseTable.Headers.index(token.lexeme)
    indexSym = parseTable.NotTerminals.index(node.symbol)

    print('indexLex: ',indexLex)
    print('indexSym: ',indexSym)

    print('token.lexeme: ',token.lexeme)
    print('node.symbol: ',node.symbol)
  
    text = parseTable.Table[indexSym+1][indexLex+1]
    
    print('text: ',text)
    text = text.split() 
    print('text: ',text)
    for i in text:
      if i in parseTable.NotTerminals: stack.append(NodeStack(i, False))
      else: stack.append(NodeStack(i, True))
        
    return stack
  else: return [node]

def PrintStack(Stack):
  symbols = Stack[0].symbol
  for i in range(1,len(Stack)):
    #symbols += ', ' + Stack[i].symbol
    symbols += Stack[i].symbol
  print('Stack: ',symbols)

def PrintTokens(Tokens):
  #symbols = '[' + Tokens[0].type + ',' + Tokens[0].lexeme + ']'
  symbols = Tokens[0].lexeme
  for i in range(1,len(Tokens)):
    symbols += Tokens[i].lexeme
  print('Tokens: ',symbols)

def Parser(stack,tokens,parseTable):
  print('')
  #parseTable.Print()
  PrintStack(stack)
  PrintTokens(tokens)
  i = 1
  while True:
    #print('\n','-- ',i,' --','\n')
    if stack[0].symbol == '$' and tokens[0].lexeme == '$':
      print(True)
      break
    if stack[0].symbol == tokens[0].lexeme and stack[0].terminal == True:
      pop = (stack.pop(0)).symbol
      pop += ' ' + (tokens.pop(0)).lexeme
      #print('                  terminal: ',pop)
    if stack[0].symbol == 'ε':
      pop = (stack.pop(0)).symbol
      #print('                  ε: ',pop)
    if stack[0].symbol != tokens[0].lexeme and stack[0].terminal == True:
      print(False)
      break
    node = stack.pop(0)  
    
    #print('node.symbol: ',node.symbol)
    stack = ReplaceStack(node,tokens[0],parseTable) + stack
    
    #PrintStack(stack)
    #PrintTokens(tokens)
    i+=1

#---------------------------
t1 = Token("lparen", "(", 1)
t2 = Token("int", "int", 1)
t3 = Token("operator", "+", 1)
t4 = Token("int", "int", 1)
t5 = Token("rparen", ")", 1)
t6 = Token("operator", "*", 1)
t7 = Token("int", "int", 1)
te = Token("$", "$", 1)

tokens = [t2,t6,t4,te]
tokens2 = [t1,t2,t3,t4,t5,te]
tokens3 = [t1,t2,t3,t4,t5,t6,t7,te]

node_E = NodeStack("E", False)
node_dolar = NodeStack("$", True)
stack = [ node_E, node_dolar ]

table = CreateTable("table.csv")
parseTable = ParseTable(table)

Parser(stack.copy(),tokens.copy(),parseTable)
Parser(stack.copy(),tokens2.copy(),parseTable)
Parser(stack.copy(),tokens3.copy(),parseTable)



# https://replit.com/@PrivateReplit/6ParserBottonUp#main.py